### PR TITLE
Bugfix converter clearing flips

### PR DESCRIPTION
**Pull request by azidar**
Bugfix: Converter was clearing flips accidentally when emitting CHIRRTL. The actual Chisel in-memory objects had the correct state, but our tracking of how "Input/Output" can clear flipped-ness was not done properly.

### Contributor Checklist

- [N/A] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [N/A] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

  - bug fix

#### API Impact

None

#### Backend Code Generation Impact

Will generate correct CHIRRTL. Unlikely that existing code relied on this buggy behavior, but if so it will now emit correctly.

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes

Bugfix: certain cases related to Compatibility mode would cause the CHIRRTL generated to be missing flips in a bundle contained in either a `Vec` or an opaque type.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chiselTests/Direction.scala [hunk 1]
```diff
@@ -379,4 +379,46 @@ class DirectionSpec extends ChiselPropSpec with Matchers with Utils {
...
+    class Decoupled extends Bundle {
+      val bits = UInt(3.W)
+      val valid = Bool()
+      val ready = Flipped(Bool())
+    }
+    class DecoupledAndMonitor extends Bundle {
+      val driver = new Decoupled()
+    }
+    class MyModule extends RawModule {
+      val w = Wire(Waivable(new Decoupled(), true, true))
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/Direction.scala:413]: what is `Waivable`?

Comment (jackkoenig) [inline src/test/scala/chiselTests/Direction.scala:413]: This mentions "opaque type" but I don't see any opaque typing here.

Comment (azidar) [reply src/test/scala/chiselTests/Direction.scala:413]: Whoops, apparently I wrote my unit test using an opaque type that I didn't move in. I'll fix this.

Comment (azidar) [reply src/test/scala/chiselTests/Direction.scala]: Replaced with `MyOpaqueType`

File: src/test/scala/chiselTests/Direction.scala [hunk 2]
```diff
@@ -379,4 +381,50 @@ class DirectionSpec extends ChiselPropSpec with Matchers with Utils {
     assert(emitted.contains("io.monitor.valid <= io.driver.valid"))
     assert(emitted.contains("io.monitor.ready <= io.driver.ready"))
   }
+  property("Bugfix: marking Vec fields with mixed directionality as Output clears inner directions") {
+    class Decoupled extends Bundle {
+      val bits = UInt(3.W)
+      val valid = Bool()
+      val ready = Flipped(Bool())
+    }
+    class DecoupledAndMonitor extends Bundle {
```
Comment (mwachs5) [inline src/test/scala/chiselTests/Direction.scala:390]: i dont get the name of this class. For clarity should this be `DecoupledMonitor` and the directon be an Input?

File: src/test/scala/chiselTests/Direction.scala [hunk 3]
```diff
@@ -379,4 +381,50 @@ class DirectionSpec extends ChiselPropSpec with Matchers with Utils {
...
+      val driver = Output(Vec(1, new Decoupled()))
+    }
+    class MyModule extends RawModule {
+      val io = IO(new DecoupledAndMonitor())
```
Comment (mwachs5) [inline src/test/scala/chiselTests/Direction.scala:394]: can you test both Input and Output case?

Comment (mwachs5) [reply src/test/scala/chiselTests/Direction.scala:394]: also can you test a case where it is not within a Bundle:

val io = IO(Input(Vec (1, new Decoupled())

File: src/test/scala/chiselTests/Direction.scala (hunk 2 above)
Comment (azidar) [reply src/test/scala/chiselTests/Direction.scala:390]: I was just using old code from a different PR, I'll rename it.

File: src/test/scala/chiselTests/Direction.scala (hunk 3 above)
Comment (azidar) [reply src/test/scala/chiselTests/Direction.scala:394]: Yup happy to :)

File: core/src/main/scala/chisel3/internal/firrtl/Converter.scala [hunk 4]
```diff
@@ -316,7 +319,7 @@ private[chisel3] object Converter {
       if (!d.opaqueType)
         fir.BundleType(d.elements.toIndexedSeq.reverse.map { case (_, e) => eltField(e) })
       else
-        extractType(d.elements.head._2, true, info)
+        extractType(d.elements.head._2, childClearDir, info)
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/firrtl/Converter.scala:322]: I missed this before but a pretty subtle bug!

**Review by jackkoenig — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**